Maintain submodule references during subgraph rewriting

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55463 Maintain submodule references during subgraph rewriting**

Differential Revision: [D27621650](https://our.internmc.facebook.com/intern/diff/D27621650)

If the `Callable` `replacement` is a Module instance in [replace_pattern](https://github.com/pytorch/pytorch/pull/55463/files#diff-14d3157db8fc26c5c1c0f18e780290a2b527c468dfc63cb3b64aa886c078502eR195), it could potentially call submodules that do not exist in the GraphModule to be transformed `gm`. The "call_module" Nodes will exist in the new Graph; they just won't reference anything concrete, which is a major error. This PR ports any used submodules to `gm` while deleting unused submodules.

Commit:
Maintain submodule references during subgraph rewriting

[ghstack-poisoned]

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -83,7 +85,114 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
+                raise RuntimeError("Unknown error type found when "
+                                   "attempting to replace missing "
+                                   "submodule in subgraph rewriter!"
+                                   "Please report a bug to PyTorch")
```
probably want to include `str(e)` in the message so we can diagnose bug reports more easily

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -83,7 +85,114 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
+                                   "Please report a bug to PyTorch")
+
+    for node in gm.graph.nodes:
+        if node.op == "call_module":
```
Would this also apply to `get_attr` nodes?

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -83,7 +85,114 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
+                                   node.target, " does not exist in the"
+                                   " original GraphModule `gm` or in "
+                                   "the replacement nn.Module "
+                                   "`replacement`")
```
When trying to understand this code I drew out this diagram:

```
                        gm has module
                      +---------------------+----------------+
                      | False               |    True        |
              +-------+---------------------+----------------+
Replacement   | False | ERROR               +  use gm mod    |
              |       |                     |  w/ type check |
has module    +-------+---------------------+----------------+
              | True  | copy w/ type check  | use gm mod     |
              +-------+---------------------+----------------+
```

Some thoughts I have are:

1) The `gm`'s submodule takes precedence over the replacement's submodule
2) Copy over the submodule from replacement only if `gm` does not have that target

I think the above two conditions could be encoded pretty simply, and the only error condition would be if neither graphmodule has the referenced target submodule.

3) I think type checking should be done elsewhere, e.g. we could simply do checks for if `call_module` truly is calling into a module in `lint`: https://github.com/pytorch/pytorch/blob/dd2bccafc599b3931555d95c222809a94d1cbe28/torch/fx/graph.py#L1010. Then we could just call `lint` after we trace the replacement callable and call lint after we're done with subgraph rewriting

jamesr66a reviewed (commented):
Cool! I think the logic could be made simpler, so added some comments about that

Commit:
Update on "Maintain submodule references during subgraph rewriting"


Differential Revision: [D27621650](https://our.internmc.facebook.com/intern/diff/D27621650)

If the `Callable` `replacement` is a Module instance in [replace_pattern](https://github.com/pytorch/pytorch/pull/55463/files#diff-14d3157db8fc26c5c1c0f18e780290a2b527c468dfc63cb3b64aa886c078502eR195), it could potentially call submodules that do not exist in the GraphModule to be transformed `gm`. The "call_module" Nodes will exist in the new Graph; they just won't reference anything concrete, which is a major error. This PR ports any used submodules to `gm` while deleting unused submodules.

[ghstack-poisoned]

Commit:
Update on "Maintain submodule references during subgraph rewriting"


Differential Revision: [D27621650](https://our.internmc.facebook.com/intern/diff/D27621650)

If the `Callable` `replacement` is a Module instance in [replace_pattern](https://github.com/pytorch/pytorch/pull/55463/files#diff-14d3157db8fc26c5c1c0f18e780290a2b527c468dfc63cb3b64aa886c078502eR195), it could potentially call submodules that do not exist in the GraphModule to be transformed `gm`. The "call_module" Nodes will exist in the new Graph; they just won't reference anything concrete, which is a major error. This PR ports any used submodules to `gm` while deleting unused submodules.

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -1010,16 +1010,28 @@ def check_arg(arg : Node, n : Optional[Node] = None) -> None:
+                                    and not isinstance(new_m_itr, torch.nn.Parameter)
+                                    and atom not in m_itr._buffers):
+                                raise RuntimeError(f'Node {node} target {node.target} {atom} of {seen_qualname} does '
+                                                    'not reference an nn.Module')
```
Hard to tell because of lint failures, but I think the error message is incorrect. The check checks if the value is a module, parameter, or buffer, but the message says `nn.Module`.


BTW, I don't know if this should apply to `get_attr`. It's valid to have a `get_attr` to a non-module/parameter/buffer, e.g. see the native interpreter example https://github.com/pytorch/examples/blob/b0649dcd638eb553238cdd994127fd40c8d9a93a/fx/native_interpreter/use_interpreter.py#L100

jamesr66a reviewed (approved):
Cool. Please fix lint!

Commit:
Update on "Maintain submodule references during subgraph rewriting"


Differential Revision: [D27621650](https://our.internmc.facebook.com/intern/diff/D27621650)

If the `Callable` `replacement` is a Module instance in [replace_pattern](https://github.com/pytorch/pytorch/pull/55463/files#diff-14d3157db8fc26c5c1c0f18e780290a2b527c468dfc63cb3b64aa886c078502eR195), it could potentially call submodules that do not exist in the GraphModule to be transformed `gm`. The "call_module" Nodes will exist in the new Graph; they just won't reference anything concrete, which is a major error. This PR ports any used submodules to `gm` while deleting unused submodules.

[ghstack-poisoned]
